src/python/pants/backend/helm/util_rules/tool.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import dataclasses
+import os
+from dataclasses import dataclass
+from typing import Iterable, Mapping
+
+from pants.backend.helm.subsystems.helm import HelmSubsystem
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.engine.environment import Environment, EnvironmentRequest
+from pants.engine.fs import CreateDigest, Digest, Directory, RemovePrefix
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.meta import frozen_after_init
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class HelmBinary:
+    path: str
+
+    env: FrozenDict[str, str]
+    immutable_input_digests: FrozenDict[str, Digest]
+
+    def __init__(
+        self,
+        path: str,
+        *,
+        helm_env: Mapping[str, str],
+        local_env: Mapping[str, str],
+        immutable_input_digests: Mapping[str, Digest],
+    ) -> None:
+        self.path = path
+        self.immutable_input_digests = FrozenDict(immutable_input_digests)
+        self.env = FrozenDict({**helm_env, **local_env})
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class HelmProcess:
+    argv: tuple[str, ...]
+    input_digest: Digest
+    description: str = dataclasses.field(compare=False)
```

Comment:
Unclear to me if uncomparable fields play nicely with the Pants engine.

cc @stuhood

Reply:
Generally they should be fine, but as expected: they will be used interchangeably, which might do odd things to logging. Unless there is a specific reason to ignore fields, I'd go ahead and include them to avoid surprises.